File: 2022/day12/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use Cwd 'abs_path';

sub parse_grid {
    my ($text) = @_;
    my @lines = split /\n/, $text;
    my @grid;
    my ($start_r, $start_c, $end_r, $end_c);

    for my $r (0 .. $#lines) {
        my @row = split //, $lines[$r];
        for my $c (0 .. $#row) {
            if ($row[$c] eq 'S') {
                ($start_r, $start_c) = ($r, $c);
                $row[$c] = 'a';
            } elsif ($row[$c] eq 'E') {
                ($end_r, $end_c) = ($r, $c);
                $row[$c] = 'z';
            }
        }
        push @grid, \@row;
    }

    return (\@grid, $start_r, $start_c, $end_r, $end_c);
}

sub bfs {
    my ($grid, $starts, $end_r, $end_c) = @_;
    my $rows = scalar @$grid;
    my $cols = scalar @{$grid->[0]};
    my %visited;
    my @queue;

    # Initialize queue with all starting positions
    for my $start (@$starts) {
        my ($r, $c) = @$start;
        push @queue, [$r, $c, 0];
        $visited{"$r,$c"} = 1;
    }

    my @directions = ([-1, 0], [1, 0], [0, -1], [0, 1]);

    while (@queue) {
        my $item = shift @queue;
        my ($r, $c, $dist) = @$item;

        # Check if we reached the end
        if ($r == $end_r && $c == $end_c) {
            return $dist;
        }

        my $current_height = ord($grid->[$r][$c]);

        for my $dir (@directions) {
            my ($dr, $dc) = @$dir;
            my ($nr, $nc) = ($r + $dr, $c + $dc);

            # Check bounds
            next if $nr < 0 || $nr >= $rows || $nc < 0 || $nc >= $cols;

            # Check if already visited
            next if $visited{"$nr,$nc"};

            my $next_height = ord($grid->[$nr][$nc]);

            # Can move if destination is at most 1 higher
            if ($next_height <= $current_height + 1) {
                $visited{"$nr,$nc"} = 1;
                push @queue, [$nr, $nc, $dist + 1];
            }
        }
    }

    return -1;  # No path found
}

sub part1 {
    my ($text) = @_;
    my ($grid, $start_r, $start_c, $end_r, $end_c) = parse_grid($text);
    return bfs($grid, [[$start_r, $start_c]], $end_r, $end_c);
}

sub part2 {
    my ($text) = @_;
    my ($grid, undef, undef, $end_r, $end_c) = parse_grid($text);

    # Find all cells with elevation 'a'
    my @starts;
    for my $r (0 .. $#$grid) {
        for my $c (0 .. $#{$grid->[$r]}) {
            if ($grid->[$r][$c] eq 'a') {
                push @starts, [$r, $c];
            }
        }
    }

    return bfs($grid, \@starts, $end_r, $end_c);
}

sub main {
    my $script_dir = dirname(abs_path(__FILE__));
    my $input_file = "$script_dir/../input.txt";

    open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
    my $text = do { local $/; <$fh> };
    close $fh;

    # Remove trailing newline if present
    chomp $text;

    print "Part 1: ", part1($text), "\n";
    print "Part 2: ", part2($text), "\n";
}

main();

File: 2022/day12/python/solution.py
#!/usr/bin/env python3
import os
from collections import deque

def parse_grid(text):
    """Parse grid and find start/end positions."""
    grid = [list(line) for line in text.strip().split('\n')]
    start = end = None

    for r, row in enumerate(grid):
        for c, ch in enumerate(row):
            if ch == 'S':
                start = (r, c)
                grid[r][c] = 'a'
            elif ch == 'E':
                end = (r, c)
                grid[r][c] = 'z'

    return grid, start, end

def bfs(grid, starts, end):
    """BFS to find shortest path from any start to end."""
    rows, cols = len(grid), len(grid[0])
    visited = set()
    queue = deque()

    for start in starts:
        queue.append((start[0], start[1], 0))
        visited.add(start)

    while queue:
        r, c, dist = queue.popleft()

        if (r, c) == end:
            return dist

        current_height = ord(grid[r][c])

        for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
            nr, nc = r + dr, c + dc

            if 0 <= nr < rows and 0 <= nc < cols and (nr, nc) not in visited:
                next_height = ord(grid[nr][nc])
                # Can move if destination is at most 1 higher
                if next_height <= current_height + 1:
                    visited.add((nr, nc))
                    queue.append((nr, nc, dist + 1))

    return -1  # No path found

def part1(text):
    """Find shortest path from S to E."""
    grid, start, end = parse_grid(text)
    return bfs(grid, [start], end)

def part2(text):
    """Find shortest path from any 'a' to E."""
    grid, _, end = parse_grid(text)

    # Find all cells with elevation 'a'
    starts = []
    for r, row in enumerate(grid):
        for c, ch in enumerate(row):
            if ch == 'a':
                starts.append((r, c))

    return bfs(grid, starts, end)

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        text = f.read()

    print('Part 1:', part1(text))
    print('Part 2:', part2(text))

if __name__ == '__main__':
    main()
